tactic/fin_cases.lean

```diff
@@ -0,0 +1,43 @@
+/-
+Copyright (c) 2018 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Scott Morrison
+
+Case bashing on `fin n`, for explicit numerals `n`.
+-/
+import tactic.linarith
+
+def fin_val {m n : ℕ} (h : m < n) : (⟨m, h⟩ : fin n).val = m := rfl
+
+namespace tactic
+open lean.parser
+open interactive interactive.types expr
+
+meta def fin_cases_at (e : expr) : tactic unit :=
+do `(fin %%n) ← infer_type e,
+   n ← eval_expr ℕ n,
+   [(_, [val, bd], _)] ← cases_core e [`val, `bd],
+   -- We now call `cases val` n times, rotating the generated goals out of the way.
+   iterate_at_most n (do val ← get_local `val, cases val, rotate 1),
+   -- We've got an absurd hypothesis, but it is messy: it looks like
+   -- `nat.succ (... (nat.succ val)) < n`
+   -- So we rewrite it as `val + 1 + ... + 1 < n`, and call `linarith` to kill it.
+   ss ← mk_const `nat.succ_eq_add_one,
+   bd ← get_local `bd,
+   (list.range n).mfoldl (λ bd _, do rewrite_hyp ss bd) bd,
+   exfalso >> `[linarith],
```

Comment:
I think this `exfalso` should be unnecessary now, it's part of `linarith`.

Reply:
I also don't think we need `linarith` here. This should be just `exact_dec_trivial`.

Reply:
I admit, I didn't read the comments a few lines up when I wrote that! It's not quite clear to me what the goal is here, maybe something like ``exact `(absurd %%bd dec_trivial)`` is needed? But you're right, `linarith` is probably overkill.

Reply:
I've removed the dependency on `linarith`.